src/main/java/beam/analysis/plots/RideHailWaitingTazAnalysis.java

```diff
@@ -0,0 +1,123 @@
+                ReserveRideHailEvent reserveRideHailEvent = (ReserveRideHailEvent) rideHailWaitingQueue.get(_personId);
+                Map<String, String> reserveRideHailEventAttributes = reserveRideHailEvent.getAttributes();
+                TAZTreeMap.TAZ pickUpLocationTAZ = beamServices.tazTreeMap().getTAZ(Double.parseDouble(reserveRideHailEventAttributes.get("startX")),
+                        Double.parseDouble(reserveRideHailEventAttributes.get("startY")));
+                double waitingTime = personEntersVehicleEvent.getTime() - reserveRideHailEvent.getTime();
+                processRideHailWaitingTimesAndTaz(reserveRideHailEvent, waitingTime,pickUpLocationTAZ);
+                // Remove the passenger from the waiting queue , as the passenger entered the vehicle.
+                rideHailWaitingQueue.remove(_personId);
+            }
+        }
+    }
+
+    /**
+     * Adds the given waiting time & TAZ to the list of cumulative waiting times per bin.
+     * @param event An occurred event
+     * @param waitingTime time spent by the passenger on waiting for the vehicle
+     * @param taz TAZ of the passenger pickup location
+     */
+    private void processRideHailWaitingTimesAndTaz(Event event, double waitingTime,TAZTreeMap.TAZ taz) {
+        //get the hour of occurrence of the event
+        int hourOfEventOccurrence = GraphsStatsAgentSimEventsListener.getEventHour(event.getTime());
+        //Add new / update the waiting time details to the list of cumulative waiting times per bin
+        Tuple<Integer, Id<TAZTreeMap.TAZ>> tuple = new Tuple<>(hourOfEventOccurrence, taz.tazId());
+        List<Double> timeList = binWaitingTimesMap.getOrDefault(tuple,new ArrayList<>());
+        timeList.add(waitingTime);
+        binWaitingTimesMap.put(tuple, timeList);
+    }
+
+    /**
+     * Write output to a csv file
+     * @param iterationNumber Current iteration number
+     * @param dataMap data to be written to the csv file
+     */
+    private void writeToCsv(int iterationNumber,Map<Tuple<Integer,Id<TAZTreeMap.TAZ>>, List<Double>> dataMap) {
+        String heading = "timeBin,TAZ,avgWait,medianWait,numberOfPickups,avgPoolingDelay,numberOfPooledPickups";
+        String fileBaseName = "rideHailWaitingStats";
+        String csvFileName = GraphsStatsAgentSimEventsListener.CONTROLLER_IO.getIterationFilename(iterationNumber, fileBaseName + ".csv");
+        BufferedWriter out = IOUtils.getBufferedWriter(csvFileName);
+        try {
+            out.write(heading);
+            out.newLine();
+            dataMap.forEach((k,v) -> {
+                DoubleSummaryStatistics stats = v.stream()
+                        .mapToDouble(x -> x)
+                        .summaryStatistics();
+                String line = k.getFirst() + "," + k.getSecond().toString() + "," + stats.getAverage() + "," +
+                        (stats.getMax() + stats.getMin())/2 + "," + 0 + "," + 0 + "," + 0;
```

Comment:
This is not the median:

> (stats.getMax() + stats.getMin())/2 

see this article to see how to do it:

http://fruzenshtein.com/scala-median-funciton/

Also, instead of making "numberOfPickups" hard coded 0, can't we just use "stats.size"?

Reply:
There is no stats.size , do you the size of the dataMap i.e., the number of entries written into the csv @colinsheppard ?